File: unity/Core/Runtime/EE/ObserverBinder.cs
namespace EE {
    public class ObserverBinder<Observer> {
        private readonly ObserverHandle _handle;
        private readonly IObserverManager<Observer> _manager;

        internal ObserverBinder(
            ObserverHandle handle,
            IObserverManager<Observer> manager) {
            _handle = handle;
            _manager = manager;
        }

        public ObserverBinder<Observer> AddObserver(Observer observer) {
            var id = _manager.AddObserver(observer);
            var manager = _manager;
            _handle.Add(() => { //
                manager.RemoveObserver(id);
            });
            return this;
        }
    }
}

File: src/ts/src/core/ObserverBinder.ts
import { IObserverManager } from "./IObserverManager";
import { ObserverHandle } from "./ObserverHandle";

export class ObserverBinder<Observer> {
    private readonly _handle: ObserverHandle;
    private readonly _manager: IObserverManager<Observer>;

    public constructor(handle: ObserverHandle, manager: IObserverManager<Observer>) {
        this._handle = handle;
        this._manager = manager;
    }

    public addObserver(observer: Observer): ObserverBinder<Observer> {
        const id = this._manager.addObserver(observer);
        const manager = this._manager;
        this._handle.add(() => {
            manager.removeObserver(id);
        });
        return this;
    }
}